modules/ServerAPI/src/server/routes/ProtocolServiceRoutes.coffee

```diff
@@ -616,3 +618,197 @@ exports.bulkPutProtocolsInternal = (protsArray, callback) ->
 			console.log error
 			callback JSON.stringify "bulk update protocols saveFailed: " + JSON.stringify error
 	)
+
+exports.getTemplateSELFile = (req, resp) ->
+	_ = require 'underscore'
+	config = require '../conf/compiled/conf.js'
+	experimentServiceRoutes = require './ExperimentServiceRoutes.js'
+	csUtilities = require '../src/javascripts/ServerAPI/CustomerSpecificServerFunctions.js'
+	request = require 'request'
+
+	protocolCode = req.body.protocolCode
+
+	exports.protocolByCodenameInternal protocolCode, (json, statusCode) ->
+		if statusCode == 200
+
+			# get the protocol metadata
+			protocolName = _.where(json.lsLabels, {lsKind: "protocol name", ignored: false})[0].labelText
+			
+			protocolMetadata = _.where(json.lsStates, {lsKind: "protocol metadata", ignored: false})[0]
+			protocolProject = _.where(protocolMetadata.lsValues, {lsKind: "project", ignored: false})[0].codeValue
+			
+			# look up the projects and convert the project code to the actual project name
+			userObject={'user':'username':req.session.passport.user.username}
+			csUtilities.getProjectsInternal userObject, (statusCode, userProjects) =>
+				if statusCode == 200
+					matchingProjects =  _.where(userProjects, {code: protocolProject, ignored: false})[0]
+					
+					# if there are matching projects, use the project name instead of the project code
+					if matchingProjects != undefined
+						protocolProject = matchingProjects.name
+
+				else
+					console.log 'got ajax error retrieving project names from project code'
+					# if we can't find the protocol name from the protocol code, we'll just use the protocol code instead...
+					# ..even though it is not ideal, we don't want to break the whole function over a small detail. 
+
+				todayDate = new Date()
+				protocolDate = todayDate.getMonth() + 1 + "/" + todayDate.getDate() + "/" + String(todayDate.getFullYear())[2..4]
+			
+				protocolScientist = req.session.passport.user.username
+				experimentServiceRoutes.experimentsByProtocolCodenameInternal protocolCode, false, (statusCode, experiments) ->
+					if statusCode == 200
+						# Part 1: Find all unique endpoints across experiments that use this protocol code
+						# arrays for recording endpoint data
+						endpointNames = []
+						endpointUnits = []
+						endpointDataTypes = []
+						endpointConcs = []
+						endpointConcUnits = []
+						endpointTimes = []
+						endpointTimeUnits = []
+						endpointHiddens = []
+
+						endpointStrings = []
+
+						for experiment in experiments
```

Comment:
We actually have the Experiment class and State classes I mentioned on the server side as well.

I think this would be simplified by using Backbone methods and our class Methods like I mentioned above:

https://github.com/mcneilco/acas/blob/7eba2bfe3ff6ec0aa039ab775cc0b045857b6ef1/modules/ServerAPI/src/server/routes/ServerUtilityFunctions.coffee#L1646

Reply:
@brianbolt I tried to implement the methods and classes on the server side too but wasn't able to get the getStatesByTypeAndKind() function to work (got ` 'Caught exception: TypeError: experiment.get(...).getStatesByTypeAndKind is not a function\n"`)...do you know how I might be able to get the function in there?

Reply:
I'd have to dig into the code a bit more than I have time for right now.  TLDR; Lets just mark this as something to possibly come back to. I don't think it's something we need to have if the code is working.

Notes:
Assuming your code snippet says something like `experiment.get('lsStates'). getStatesByTypeAndKind`. If `getStatesByTypeAndKind` is not a function, then it might not have been set as a [StateList](https://github.com/mcneilco/acas/blob/040800fdb4b7b7ec4405d5930ba61b7e16219801/modules/ServerAPI/src/server/routes/ServerUtilityFunctions.coffee#L756) class.  I might just be a plain dict object.  I would expect though if `experiment` had been created as an instance of [Experiment](https://github.com/mcneilco/acas/blob/040800fdb4b7b7ec4405d5930ba61b7e16219801/modules/ServerAPI/src/server/routes/ServerUtilityFunctions.coffee#L1646) then it would have a [StateList](https://github.com/mcneilco/acas/blob/040800fdb4b7b7ec4405d5930ba61b7e16219801/modules/ServerAPI/src/server/routes/ServerUtilityFunctions.coffee#L1657) property

Reply:
Yep, I was using it `experiment.get('lsStates').getStatesByTypeAndKind`, with a line before it explicitly calling `experiment = new Experiment experiment ` and it was throwing the error. 

After your comment I also tried to also break it up with `experimentLsStates = new StateList experiment.get('lsStates')` and then calling `experimentLsStates.getStatesByTypeAndKind` which still threw the same error?